File: BSTLCA.java
// Time Complexity : O(h)
// Space Complexity : O (h)
// Did this code successfully run on Leetcode :
// Any problem you faced while coding this :


// Your code here along with comments explaining your approach
/*
 * We traverse the entire tree based on which side of the tree the values could be present on.
 * 1 - If both p and q are greater than q, then the roots lie on the right side.
 * 2 - If both p and q are lesser than q, then the roots lie on the left side.
 * 3 - Else if either p or q is equal to the root, then return the root simply as the LCA
 */

public class BSTLCA {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        if(p.val > root.val && q.val > root.val)
        {
            return lowestCommonAncestor(root.right,p,q);
        }
        else if(p.val < root.val && q.val < root.val)
        {
            return lowestCommonAncestor(root.left,p,q);
        }
        else
            return root;
    }
}

File: BST_LCA.py
 """Time Complexity : O(n)
 Space Complexity : 
 Did this code successfully run on Leetcode : yes
 Any problem you faced while coding this : yes, initially i just returned the root, but due to this the main return was of a null value


 Your code here along with comments explaining your approach """

 class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        print(root.val)
        if p.val > root.val and q.val> root.val:
            return self.lowestCommonAncestor(root.right, p, q)
        elif q.val < root.val and p.val<root.val:
            return self.lowestCommonAncestor(root.left, p, q)
        else:
            return root
            
        
